clients/drcachesim/tracer/output.cpp

```diff
@@ -916,22 +914,51 @@ process_and_output_buffer(void *drcontext, bool skip_size_cap)
         open_new_thread_file(drcontext, get_local_window(data));
     }
 
+    size_t header_size = buf_hdr_slots_size;
+    // For online we already wrote the thread header but for offline it is in
+    // the first buffer.
+    if (data->has_thread_header && op_offline.get_value())
+        header_size += data->init_header_size;
+    data->has_thread_header = false;
+
+    if (align_attach_detach_endpoints()) {
+        // This is the attach counterpart to instru_t::set_frozen_timestamp(): we place
+        // timestamps at buffer creation, but that can be before we're fully attached.
+        // We update any too-early timestamps to reflect when we actually started
+        // tracing.  (Switching back to timestamps at buffer output is actually
+        // worse as we then have the identical frozen timestamp for all the flushes
+        // during detach, plus they are all on the same cpu too.)
+        size_t stamp_offs =
+            header_size > buf_hdr_slots_size ? header_size - buf_hdr_slots_size : 0;
+        uint64 min_timestamp = attached_timestamp.load(std::memory_order_acquire);
+        bool found = instru->refresh_unit_header_timestamp(data->buf_base + stamp_offs,
```

Comment:
In normal operation, we don't expect the stamp_offs to NOT be the timestamp entry. So maybe it's better to assert found inside refresh_unit_header_timestamp.

Reply:
My logic was treating it like a library routine but it is pretty integrated so having this a pre-req and assert seems reasonable.  Done.